### PR TITLE
lib: added parentheses verifying the lack of ambiguities

### PR DESCRIPTION
**Pull request by DeHess**
Fix coding guideline MISRA C:2012 Rule 12.1 in lib:

> The precedence of operators within expressions should be made explicit.

This PR is part of the enhancement issue https://github.com/zephyrproject-rtos/zephyr/issues/48002 which port the coding guideline fixes done by BUGSENG on the https://github.com/zephyrproject-rtos/zephyr/tree/v2.7-auditable-branch back to main

The commit in this PR is a subset of the original auditable-branch commit:
https://github.com/zephyrproject-rtos/zephyr/commit/29155bdd6c8f8c2c0bdf0562191b4b968b3a818a

### CODE REVIEW

File: lib/libc/minimal/source/stdlib/atoi.c
Comment (ycsin) [inline lib/libc/minimal/source/stdlib/atoi.c:56]:
```suggestion
		n = (10 * n) - (*s++ - '0');
```

**Review by ycsin — DISMISSED** [review]
LGTM, some pre-existing issue nits that could use some fixing.

**Comment by keith-packard** [comment]
I'm usually in favor of extra parens so we don't have to remember all of the C precedence levels, but do we really need parens around (a * b) + c too? Or (a + b) >= c? At some point, it does get a bit chatty.

**Review by nashif — DISMISSED** [review]

**Review by cfriedt — DISMISSED** [review]

**Review by stephanosio — DISMISSED** [review]

**Review by simhein — DISMISSED** [review]

**Comment by simhein** [comment]
> I'm usually in favor of extra parens so we don't have to remember all of the C precedence levels, but do we really need parens around (a * b) + c too? Or (a + b) >= c? At some point, it does get a bit chatty.

@keith-packard Thanks for your Feedback. I generally agree that this gets chatty in some way. On the other hand I don't know a way to partially use this guideline to be honest. If we check that guideline with a tooling, than the tooling gets very chatty and complain about it every time, to solve this we need to add something in the code which makes the code not that readable which I want to avoid. I will keep that topic in mind and look into it to find a way how we can maybe handle such things in future.

**Review by simhein — DISMISSED** [review]

**Review by ycsin — DISMISSED** [review]

**Review by aescolar — DISMISSED** [review]

File: lib/heap/heap.h
Comment (fabiobaltieri) [inline lib/heap/heap.h:247]: Again, this is the exact counter example in https://wiki.sei.cmu.edu/confluence/display/c/EXP00-C.+Use+parentheses+for+precedence+of+operation

**Review by fabiobaltieri — CHANGES_REQUESTED** [review]

**Comment by DeHess** [comment]
@fabiobaltieri 
Thanks for the input. You are right, I have removed the offending parentheses.

In my defense, I mostly just port forward the changes made by Bugseng, even when I don't agree with the coding style. In hindsight, my blind trust in their knowledge of the coding guidelines might have been misplaced- I just assumed they knew better. 

It seems they applied their own interpretations of the rules / coding style.

**Review by ycsin — DISMISSED** [review]

**Review by aescolar — DISMISSED** [review]

File: lib/heap/heap.c
Comment (fabiobaltieri) [inline lib/heap/heap.c:433]: this is also unnecessary, there's no operator ambiguity already

File: lib/utils/timeutil.c
Comment (fabiobaltieri) [inline lib/utils/timeutil.c:42]: same

**Review by fabiobaltieri — COMMENTED** [review]
@DeHess that's fine just trying to understand what and why and how are we doing this, nothing personal, no hard feelings: ~~are these discovered using some auditing tools or by manually checking the code?~~ oh, Bugseng, I see, so where's the discontinuity? Am I reading the rules wrong? Is it intentionally flagging basic algebra priority as issues? It seems a bit ridiculous to be perfectly honest.

To state the obvious I'm very much not a fan of these, on the basis of:
- the rule in our own documentation is "advisory"
- the CERT C page only refers to bitwise operators for these, which I very much agree with and I think that GCC has a warning for these already, this "fixes" a lot of logical operator, who's precedence is **very** well understood

Anyway seems like there's consensus that these needs to be "fixed" for whatever reasons so let's go with it but I spotted another couple of redundant ones.

File: lib/heap/heap_info.c
Comment (cfriedt) [inline lib/heap/heap_info.c:70]: This is kind of funny, because the tool is just enforcing spacing but as a result, it misses the operator precedence "violation" (i.e. the tool is inconsistent).

**Comment by simhein** [comment]
> @DeHess that's fine just trying to understand what and why and how are we doing this, nothing personal, no hard feelings: ~are these discovered using some auditing tools or by manually checking the code?~ oh, Bugseng, I see, so where's the discontinuity? Am I reading the rules wrong? Is it intentionally flagging basic algebra priority as issues? It seems a bit ridiculous to be perfectly honest.
> 
> To state the obvious I'm very much not a fan of these, on the basis of:
> 
>     * the rule in our own documentation is "advisory"
> 
>     * the CERT C page only refers to bitwise operators for these, which I very much agree with and I think that GCC has a warning for these already, this "fixes" a lot of logical operator, who's precedence is **very** well understood
> 
> 
> Anyway seems like there's consensus that these needs to be "fixed" for whatever reasons so let's go with it but I spotted another couple of redundant ones.

@fabiobaltieri Maybe a little bit more context for these kind of PRs to understand the origin of them better:  Bugseng was contracted from the project to fix the guidelines in the sources, they used their tool ECLAIR for it which we are evaluation as well as checker for the project itself. They did the fixes on a separate branch a while ago and we try to merge them now to the main if we don't merge theme, the contract would have been a waste of money in my opinion.

And thank you for flagging all these PRs, I think it is valuable to get another opinion on it and re-think some rules.

About the "advisory" and severity in the documentation, I think it is a little bit misleading tbh. The severity belongs to the Misra rule itself and our guidelines are a subset of these rules. As far es I understood the discussions/decision about the rules is that these guidelines are set for the project but I will double check it and update the documentation to be more clear and specific!

File: lib/heap/heap_info.c
Comment (simhein) [reply lib/heap/heap_info.c:70]: True this needs some investigation.

**Review by simhein — CHANGES_REQUESTED** [review]

**Comment by fabiobaltieri** [comment]
Sounds good, thanks for explaining @simhein. Frankly with how fast the project moves and how many contributors and new contributors we have, I think that any code maintenance efforts without a solid plan on how to keep the code base into compliance is a bit of an exercise in futility. Bare minimum the tools should be accessible to all developers, ideally integrated in CI, ideally open source. But hey let's try I guess, sure it's all experience, let's just make sure we at least identify if the tools is doing what it's supposed to do or not.

File: lib/heap/heap_info.c
Comment (DeHess) [reply lib/heap/heap_info.c:70]: Yeah that's my bad, I split it off into a seperate commit for clarity.

**Review by fabiobaltieri — APPROVED** [review]

**Review by simhein — APPROVED** [review]

**Review by cfriedt — APPROVED** [review]

**Comment by romkell** [comment]
> > @DeHess that's fine just trying to understand what and why and how are we doing this, nothing personal, no hard feelings: ~are these discovered using some auditing tools or by manually checking the code?~ oh, Bugseng, I see, so where's the discontinuity? Am I reading the rules wrong? Is it intentionally flagging basic algebra priority as issues? It seems a bit ridiculous to be perfectly honest.
> > To state the obvious I'm very much not a fan of these, on the basis of:
> > ```
> > * the rule in our own documentation is "advisory"
> > 
> > * the CERT C page only refers to bitwise operators for these, which I very much agree with and I think that GCC has a warning for these already, this "fixes" a lot of logical operator, who's precedence is **very** well understood
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Anyway seems like there's consensus that these needs to be "fixed" for whatever reasons so let's go with it but I spotted another couple of redundant ones.
> 
> @fabiobaltieri Maybe a little bit more context for these kind of PRs to understand the origin of them better: Bugseng was contracted from the project to fix the guidelines in the sources, they used their tool ECLAIR for it which we are evaluation as well as checker for the project itself. They did the fixes on a separate branch a while ago and we try to merge them now to the main if we don't merge theme, the contract would have been a waste of money in my opinion.
> 
> And thank you for flagging all these PRs, I think it is valuable to get another opinion on it and re-think some rules.
> 
> About the "advisory" and severity in the documentation, I think it is a little bit misleading tbh. The severity belongs to the Misra rule itself and our guidelines are a subset of these rules. As far es I understood the discussions/decision about the rules is that these guidelines are set for the project but I will double check it and update the documentation to be more clear and specific!

Basically in response to all:
- The MISRA 12.1 is very clear on what should be done. It is just not accessible through the project guidelines due to licensing restrictions - an open MISRA C 2012 spec would be nice to have in a project public form which is rather hard to do. (How to make it public to the project without making it public to the world? A closed repo with controlled access for maintainer and assigned contributors?).
- MISRAs own (addition) examples are only half the story.
- Our own guidelines are ambiguous - trying to apply Cert C EXP00-C and MISRA 12.1 is bound to fail. MISRA seems more strict that Cert C. A tool, ECLAIR or which ever, will follow the rules and only be able to check for one or the other. Applying both can work, but might also throw violations. And giving both in our guidelines leads to these situations where people take what they like better.
- We need to decide for one or the other. Staying with both is confusing. If we drop MISRA 12.1 in favor of CERT EXP00-C we might have to argue with a certification body how we mitigate the gap between the two, since I feel MISRA is stricter. If CERT EXP00-C without comment, fine.
- MISRAs rational is, that there are too many operator precedence levels (16) in C language and therefore not intuitive, so that it can lead less experienced programmers to make mistakes. Further the idea is that others, e.g. reviewer and maintainer immediately recognize the programmers idea since it was made explicit: `(x * y) + z` vs `x * (y + z)` even if mathematically the parentheses are not required.
- In the end I would be pragmatic about it: we can possibly argue with the certification body and possibly find mitigations for the gaps or type some extra parentheses. And if the expression becomes unreadable due to the parentheses it probably was already hard to understand before and could (should?) have been split.

**Comment by romkell** [comment]
> > @DeHess that's fine just trying to understand what and why and how are we doing this, nothing personal, no hard feelings: ~are these discovered using some auditing tools or by manually checking the code?~ oh, Bugseng, I see, so where's the discontinuity? Am I reading the rules wrong? Is it intentionally flagging basic algebra priority as issues? It seems a bit ridiculous to be perfectly honest.
> > To state the obvious I'm very much not a fan of these, on the basis of:
> > ```
> > * the rule in our own documentation is "advisory"
> > 
> > * the CERT C page only refers to bitwise operators for these, which I very much agree with and I think that GCC has a warning for these already, this "fixes" a lot of logical operator, who's precedence is **very** well understood
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Anyway seems like there's consensus that these needs to be "fixed" for whatever reasons so let's go with it but I spotted another couple of redundant ones.
> 
> @fabiobaltieri Maybe a little bit more context for these kind of PRs to understand the origin of them better: Bugseng was contracted from the project to fix the guidelines in the sources, they used their tool ECLAIR for it which we are evaluation as well as checker for the project itself. They did the fixes on a separate branch a while ago and we try to merge them now to the main if we don't merge theme, the contract would have been a waste of money in my opinion.
> 
> And thank you for flagging all these PRs, I think it is valuable to get another opinion on it and re-think some rules.
> 
> About the "advisory" and severity in the documentation, I think it is a little bit misleading tbh. The severity belongs to the Misra rule itself and our guidelines are a subset of these rules. As far es I understood the discussions/decision about the rules is that these guidelines are set for the project but I will double check it and update the documentation to be more clear and specific!

Basically in response to all:
- The MISRA 12.1 is very clear on what should be done. It is just not accessible through the project guidelines due to licensing restrictions - an open MISRA C 2012 spec would be nice to have in a project public form which is rather hard to do. (How to make it public to the project without making it public to the world? A closed repo with controlled access for maintainer and assigned contributors?).
- MISRAs own (addition) examples are only half the story.
- Our own guidelines are ambiguous - trying to apply Cert C EXP00-C and MISRA 12.1 is bound to fail. MISRA seems more strict that Cert C. A tool, ECLAIR or which ever, will follow the rules and only be able to check for one or the other. Applying both can work, but might also throw violations. And giving both in our guidelines leads to these situations where people take what they like better.
- We need to decide for one or the other. Staying with both is confusing. If we drop MISRA 12.1 in favor of CERT EXP00-C we might have to argue with a certification body how we mitigate the gap between the two, since I feel MISRA is stricter. If CERT EXP00-C without comment, fine.
- MISRAs rational is, that there are too many operator precedence levels (16) in C language and therefore not intuitive, so that it can lead less experienced programmers to make mistakes. Further the idea is that others, e.g. reviewer and maintainer immediately recognize the programmers idea since it was made explicit: `(x * y) + z` vs `x * (y + z)` even if mathematically the parentheses are not required.
- In the end I would be pragmatic about it: we can possibly argue with the certification body and possibly find mitigations for the gaps or type some extra parentheses. And if the expression becomes unreadable due to the parentheses it probably was already hard to understand before and could (should?) have been split.

> @romkell thanks for the detail response, how we think we could keep any part of an open source project in compliance with a close standard, who's details are super secret and compliance tooling unavailable is very much beyond my understanding. But thanks for highlighting the incoherencies with the process, it seems to me like this needs some more discussion.
> 
> Just one point I'd like an explicit clarification, since like most people I don't (and frankly, I don't a particularly want to) have access to the standard:
> 
> > * MISRAs rational is, that there are too many operator precedence levels (16) in C language and therefore not intuitive, so that it can lead less experienced programmers to make mistakes. Further the idea is that others, e.g. reviewer and maintainer immediately recognize the programmers idea since it was made explicit: `(x * y) + z` vs `x * (y + z)` even if mathematically the parentheses are not required.
> 
> This is explicitly listed? ANY operator priority should be explicit? Including addition/subtraction versus multiplcation/division?

@fabiobaltieri :No, they do not have every combination as an example in the spec. This one was made up by me to emphasis the programmers intent.
They describe rules how to determine when to set the parentheses.

The example they show is `x << y + z` where `y + z` has to be in parentheses even though not required, unless bit shift `x << y` was meant to be done first, hence: `x << y + z` = `x << (y + z)` != `(x << y) + z`. I guess most people would do `x << (y + z)` naturally. At least I would.

The rule here is that one has to determine the precedence of the expression which is the root operand or operator. Expressions that have a precedence from 2 to 12 is required to have parentheses around any operand which has precedence less than 13 AND which precedence is greater the expressions precedence.

Close to not understandable with precedence table, but without probably all Greek. That is, I guess why you want to have a tool for that.
Is all of that beneficial in all situations - I do not know. I guess not. But inventing our own rules will mean that you will have to implement that into a checker tool, because from the shelf we get MISRA C or CERT C or ?? and get our rules accepted by certification authorities. Are some extra parentheses that evil? Not to me.

**Comment by fabiobaltieri** [comment]
@romkell ok thanks for explaining, I don't think this is enforcable without tooling.

> The example they show is `x << y + z` where `y + z` has to be in parentheses even though not required, unless bit shift `x << y` was meant to be done first, hence: `x << y + z` = `x << (y + z)` != `(x << y) + z`. I guess most people would do `x << (y + z)` naturally. At least I would.

Yes that's a good example. No people wont't do it naturally. Yes we already have tooling to catch that case, GCC has a warning, we fail on GCC warnings, we try to build all code in CI:

```
/usr/local/google/home/fabiobaltieri/zephyrproject/zephyr/samples/basic/blinky/src/main.c:37:24: warning: suggest parentheses around ‘+’ inside ‘<<’ [-Wparentheses]
   37 |         printf("%d", 1 << 2 + 3);
      |                        ^~
```

**Comment by fabiobaltieri** [comment]
@romkell yes the whole applicability of these rules deserves its own discussion and should be funnelled to the appropriate channel, probably process WG as a start.

The whole parenthesis details is just for my own knowlege since I see these PR being thrown my way and I'm not happy with stamping them without understanding what's behind them properly. So as annoying as it sounds (sorry for that, I'm really having a hard time grasping the rule):

> Same precedence order `+ before <<` and `* before +`. Misra 12.1 would/should through a violation also for the second.

So MISRA would require to use parenthesis for basic arithmetic priority? `1 + 2 * 3` would be NOT ok?

Again sorry for insisting on that details I'm just trying to understand what we are dealing with.

**Comment by romkell** [comment]
> @romkell yes the whole applicability of these rules deserves its own discussion and should be funnelled to the appropriate channel, probably process WG as a start.
> 
> The whole parenthesis details is just for my own knowlege since I see these PR being thrown my way and I'm not happy with stamping them without understanding what's behind them properly. So as annoying as it sounds (sorry for that, I'm really having a hard time grasping the rule):
> 
> > Same precedence order `+ before <<` and `* before +`. Misra 12.1 would/should through a violation also for the second.
> 
> So MISRA would require to use parenthesis for basic arithmetic priority? i.e. `1 + 2 * 3` would be NOT ok?
> 
> Again sorry for insisting on that details I'm just trying to understand what we are dealing with.

I have not checked `1 + 2 * 3` against Misra 12.1 with a commercial SCA. But the 12.1 rules do not mention basic arithmetic prios. It is all related to C's precedence which include basic arithmetic prios but does not treat them separately.

Here the result of CppCheck (OSS SCA tool)

```
	(void)printf("%d", 1 << 2 + 3);
	(void)printf("%d", 1 + 2 * 3);
	(void)printf("%d", 2 * 3 + 1 );
```

```
/home/user/workspace/esf20project/esf20/tests/middleware/modbus/itest/src/modbus_srv_hr_cbs.c:45:23: style: Operands shall not be of an inappropriate essential type. [misra-c2012-10.1]
 (void)printf("%d", 1 << 2 + 3);
                      ^
/home/user/workspace/esf20project/esf20/tests/middleware/modbus/itest/src/modbus_srv_hr_cbs.c:45:23: style: The precedence of operators within expressions should be made explicit. [misra-c2012-12.1]
 (void)printf("%d", 1 << 2 + 3);
                      ^
/home/user/workspace/esf20project/esf20/tests/middleware/modbus/itest/src/modbus_srv_hr_cbs.c:46:23: style: The precedence of operators within expressions should be made explicit. [misra-c2012-12.1]
 (void)printf("%d", 1 + 2 * 3);
                      ^
/home/user/workspace/esf20project/esf20/tests/middleware/modbus/itest/src/modbus_srv_hr_cbs.c:47:27: style: The precedence of operators within expressions should be made explicit. [misra-c2012-12.1]
 (void)printf("%d", 2 * 3 + 1 );
                          ^
```
> So MISRA would require to use parenthesis for basic arithmetic priority? i.e. `1 + 2 * 3` would be NOT ok?

Believing that CppCheck is close enough to a commercial tool, the short answer is: yes.

**Pull request merged**
